source/GameInterface/Services/Clans/Patches/ClanAddCompanionPatch.cs

```diff
@@ -13,34 +13,28 @@ namespace GameInterface.Services.Clans.Patches
     [HarmonyPatch(typeof(AddCompanionAction), "ApplyInternal")]
     public class ClanAddCompanionPatch
     {
-        private static readonly AllowedInstance<Hero> AllowedInstance = new AllowedInstance<Hero>();
-
         static bool Prefix(Clan clan, Hero companion)
         {
-            if (AllowedInstance.IsAllowed(companion)) return true;
+            if (AllowedThread.IsThisThreadAllowed()) return true;
 
             if (PolicyProvider.AllowOriginalCalls) return true;
 
             if (ModInformation.IsClient && clan != Clan.PlayerClan) return false;
 
-            CallStackValidator.Validate(companion, AllowedInstance);
-
             MessageBroker.Instance.Publish(clan, new CompanionAdded(clan.StringId, companion.StringId));
 
             return false;
         }
 
         public static void RunOriginalAddCompanion(Clan clan, Hero companion)
         {
-            using (AllowedInstance)
+            GameLoopRunner.RunOnMainThread(() =>
             {
-                AllowedInstance.Instance = companion;
-
-                GameLoopRunner.RunOnMainThread(() =>
+                using(new AllowedThread())
                 {
                     AddCompanionAction.Apply(clan, companion);
-                }, true);
-            }
+                }
+            }, true);
```

Comment:
Can you update this to non-blocking in your branch so we dont forget
```suggestion
            });
```